hnn_core/optimization.py

```diff
@@ -447,14 +583,17 @@ def _myoptrun(new_params):
         # tiny negative weights are possible. Clip them to 0.
         opt_results[opt_results < 0] = 0
 
-        # update opt_params for the next round
-        for var_name, value in zip(opt_params['ranges'], opt_results):
-            opt_params['ranges'][var_name]['initial'] = value
+        # update opt_params for the next round if total rmse decreased
+        avg_rmse = _rmse(opt_dpls['best_dpl'],
+                         opt_dpls['target_dpl'],
+                         tstop=tstop,
+                         weights=None)
+        if avg_rmse <= best_rmse:
+            best_rmse = avg_rmse
+            for var_name, value in zip(opt_params['ranges'], opt_results):
+                opt_params['ranges'][var_name]['initial'] = value
 
-    # save the optimized params
-    for var_name in opt_params['ranges']:
-        params[var_name] = opt_params['ranges'][var_name]['initial']
+            net = net_opt
```

Comment:
Correct. And it's weighted to prioritize different time points differently within that segment. I suspect this may be the root cause of #466 (note that I observed #466 on the current master as well as my fix_optimization branch), but more work is required to flush this out.